File: Leetcode/tree/1022_SumOfRootToLeafBinaryNumbers/Solution.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public int sumRootToLeaf(TreeNode root) {
        int[] res = new int[1];
        dfs(root, 0, res);
        return res[0];
    }
        private void dfs(TreeNode root, int val, int[] res){
            if(root==null) return;
            val = val*2 + root.val;
            dfs(root.left, val, res);
            dfs(root.right, val, res);
            if(root.left==null && root.right==null){
                res[0]+= val;
                return;
            }
        }
}

File: Leetcode/tree/1022_SumOfRootToLeafBinaryNumbers/Solution.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
private:
    void dfs(TreeNode* root, int val, int &res){
        if(root==nullptr) return;
        val = val*2+root->val;
        dfs(root->left, val, res);
        dfs(root->right, val, res);
        if(root->left==nullptr && root->right==nullptr){
            res += val;
            return;
        }
    }
public:
    int sumRootToLeaf(TreeNode* root) {
        int res = 0;
        dfs(root, 0, res);
        return res;
    }
};
